### PR TITLE
Add data masking client, test, recording

### PR DESCRIPTION
**Pull request by nathannfan**
## Description
Add generated client, test and recording for Data Masking APIs.

TODO: need to update AssemblyInfo.cs because we already have 2 SQL SDK pull requests open that haven't been reviewed yet, so need to determine the version based on which ones merge first.

https://github.com/Azure/azure-rest-api-specs/pull/1321
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by nathannfan** [comment]
Corresponding swagger pull request: https://github.com/Azure/azure-rest-api-specs/pull/1321

Should be merged soon, is only blocked because the reviewer assigned is OOF.

**Comment by nathannfan** [comment]
I don't believe the validation build is correct.  I made a one line change that shouldn't be breaking anything, as it was correcting a mistaken value.

File: src/SDKs/SqlManagement/Sql.Tests/Utilities/SqlManagementTestUtilities.cs [hunk 1]
```diff
@@ -40,6 +40,10 @@ public class SqlManagementTestUtilities
 
         public const string DefaultEuapPrimaryLocation = "East US 2 EUAP";
 
+        public const string DefaultLogin = "dummylogin";
+
+        public const string DefaultPassword = "Un53cuRE!";
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Sql.Tests/Utilities/SqlManagementTestUtilities.cs]: can we generate it in runtime with some random seed?

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+using Microsoft.Azure.Test.HttpRecorder;
+
+namespace Sql.Tests
+{
+    public class DataMaskingScenarioTests
+    {
+        [Fact]
+        public void TestCreateUpdateGetDataMaskingRules()
+        {
+            string testPrefix = "sqlcrudtest-";
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: Please add data masking string to the test prefix

File: src/SDKs/SqlManagement/Management.Sql/generate.cmd [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 @echo off
 if  "%1" == "" (
-    set specFile="https://raw.githubusercontent.com/Azure/azure-rest-api-specs/9346f0adc0caddbfa33dbe69a469c199705cf38a/arm-sql/compositeSql.json"
+    set specFile="https://raw.githubusercontent.com/Azure/azure-rest-api-specs/03ff383aac77901093d96c049e248d2c7a3cf4ff/arm-sql/compositeSql.json"
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Management.Sql/generate.cmd]: why do we need this change?

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/TransparentDataEncryptionStatus.cs [hunk 4]
```diff
@@ -27,4 +27,32 @@ public enum TransparentDataEncryptionStatus
         [EnumMember(Value = "Disabled")]
         Disabled
     }
+    internal static class TransparentDataEncryptionStatusEnumExtension
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/TransparentDataEncryptionStatus.cs]: how is this change related?

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyUseServerDefault.cs [hunk 5]
```diff
@@ -27,4 +27,32 @@ public enum SecurityAlertPolicyUseServerDefault
...
+        }
+
+        internal static SecurityAlertPolicyUseServerDefault? ParseSecurityAlertPolicyUseServerDefault( this string value )
+        {
+            switch( value )
+            {
+                case "Enabled":
+                    return SecurityAlertPolicyUseServerDefault.Enabled;
+                case "Disabled":
+                    return SecurityAlertPolicyUseServerDefault.Disabled;            }
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyUseServerDefault.cs]: the "}" should be in a new line

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyState.cs [hunk 6]
```diff
@@ -29,4 +29,36 @@ public enum SecurityAlertPolicyState
...
+        internal static SecurityAlertPolicyState? ParseSecurityAlertPolicyState( this string value )
+        {
+            switch( value )
+            {
+                case "New":
+                    return SecurityAlertPolicyState.New;
+                case "Enabled":
+                    return SecurityAlertPolicyState.Enabled;
+                case "Disabled":
+                    return SecurityAlertPolicyState.Disabled;            }
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyState.cs]: the "}" should be in a new line

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        /// maskingFunction, and specify ruleState as disabled. However, if the
+        /// rule doesn't already exist, the rule will be created with ruleState
+        /// set to enabled, regardless of the provided value of ruleState.
+        /// Possible values include: 'Disabled', 'Enabled'
+        /// </summary>
+        [JsonProperty(PropertyName = "properties.ruleState")]
+        public DataMaskingRuleState? RuleState { get; set; }
+
+        /// <summary>
+        /// Gets or sets the schema name on which the data masking rule is
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs]: can the comment be in one line? (here and below)

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+        /// <summary>
+        /// Validate the object.
+        /// </summary>
+        /// <exception cref="ValidationException">
+        /// Thrown if validation fails
+        /// </exception>
+        public virtual void Validate()
+        {
+            if (SchemaName == null)
```
Comment (yaakoviyun) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs]: data masking function should not be null too...

**Review by yaakoviyun — APPROVED** [review]

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                });
+
+                // Create test table with columns
+                HttpRecorderMode testMode = HttpMockServer.GetCurrentMode();
+
+                if (testMode != HttpRecorderMode.Playback)
+                {
+                    SqlConnectionStringBuilder builder = new SqlConnectionStringBuilder()
+                    {
+                        DataSource = string.Format("{0}.database.windows.net", server.Name),
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: I think there is `server.FullyQualifiedDomainName` that you can use

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                        UserID = SqlManagementTestUtilities.DefaultLogin,
+                        Password = SqlManagementTestUtilities.DefaultPassword,
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: Nice change, can you also update ImportExport tests to use these?

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+                sqlClient.FirewallRules.CreateOrUpdate(resourceGroup.Name, server.Name, "sqltestrule", new FirewallRule()
+                {
+                    StartIpAddress = "0.0.0.0",
+                    EndIpAddress = "255.255.255.255"
+                });
+
+                // Create test table with columns
+                HttpRecorderMode testMode = HttpMockServer.GetCurrentMode();
+
+                if (testMode != HttpRecorderMode.Playback)
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: Please add comment explaining why it's not needed in Playback, i.e. because there is no HTTP requests in this block and it requires the server to really exist

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+namespace Sql.Tests
+{
+    public class DataMaskingScenarioTests
+    {
+        [Fact]
+        public void TestCreateUpdateGetDataMaskingRules()
+        {
+            string testPrefix = "sqlcrudtest-";
+            string suiteName = this.GetType().FullName;
+            SqlManagementTestUtilities.RunTestInNewV12Server(suiteName, "TestCreateUpdateGetDataMaskingRules", testPrefix, (resClient, sqlClient, resourceGroup, server) =>
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: Can you use new SqlManagementTestContext instead? See Restore LTR test for example

**Review by jaredmoo — CHANGES_REQUESTED** [review]
Looks great, just minor comments

File: src/SDKs/SqlManagement/Sql.Tests/Utilities/SqlManagementTestUtilities.cs (hunk 1 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Sql.Tests/Utilities/SqlManagementTestUtilities.cs]: This is how we've been doing tests for all the SDKs, so I think it should be fixed separate from this PR.  Please create a separate github issue for this.

File: src/SDKs/SqlManagement/Management.Sql/generate.cmd (hunk 3 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Management.Sql/generate.cmd]: Ask from the .net SDK team to know what version of compositeSql.json is used to generate client

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs (hunk 2 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: okay.

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/TransparentDataEncryptionStatus.cs (hunk 4 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/TransparentDataEncryptionStatus.cs]: Generated client from latest swagger

File: src/SDKs/SqlManagement/Sql.Tests/Utilities/SqlManagementTestUtilities.cs (hunk 1 above)
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/Utilities/SqlManagementTestUtilities.cs]: It's [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines) number 2 :)

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyState.cs (hunk 6 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyState.cs]: This is generated code and changing the formatting would just be overwritten the next time someone generates client

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs (hunk 7 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs]: This is generated code and changing the formatting would just be overwritten the next time someone generates client

File: src/SDKs/SqlManagement/Management.Sql/generate.cmd (hunk 3 above)
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Management.Sql/generate.cmd]: It's [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines) number 2 :)

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs (hunk 8 above)
Comment (nathannfan) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/DataMaskingRule.cs]: Data masking function is stored here as a non-nullable enum

File: src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyUseServerDefault.cs (hunk 5 above)
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Management.Sql/Generated/Models/SecurityAlertPolicyUseServerDefault.cs]: Filed https://github.com/Azure/autorest/issues/2386 against autorest code generator to fix this

File: src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+namespace Sql.Tests
+{
+    public class DataMaskingScenarioTests
+    {
+        [Fact]
+        public void TestCreateUpdateGetDataMaskingRules()
+        {
+            string testPrefix = "sqldatamaskingcrudtest-";
+            string suiteName = this.GetType().FullName;
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/DataMaskingScenarioTests.cs]: `suiteName` unused?

**Review by jaredmoo — APPROVED** [review]

**Review by shahabhijeet — APPROVED** [review]

**Comment by nathannfan** [comment]
@shahabhijeet Thanks, abhijeet!
